micro-rdk/src/common/data_manager.rs

```diff
@@ -262,43 +252,139 @@ where
+                        continue;
+                    }
+                    if next_message.is_empty()
+                        || ((next_message.len() + current_chunk_size) > MAX_SENSOR_CONTENTS_SIZE)
+                    {
+                        let data: Result<Vec<SensorData>, prost::DecodeError> =
+                            current_chunk.drain(..).map(SensorData::decode).collect();
+                        let data = match data {
+                            Ok(data) => data,
+                            Err(err) => {
+                                log::error!(
+                                    "error decoding readings for collector key ({:?}): {:?}",
+                                    collector_key,
+                                    err
+                                );
+                                break false;
+                            }
+                        };
+                        let upload_request = DataCaptureUploadRequest {
+                            metadata: Some(UploadMetadata {
+                                part_id: self.part_id.clone(),
+                                component_type: collector_key.component_type.clone(),
+                                r#type: DataType::TabularSensor.into(),
+                                component_name: collector_key.r_name.clone(),
+                                method_name: collector_key.method.to_string(),
+                                ..Default::default()
+                            }),
+                            sensor_contents: data,
+                        };
+                        match app_client
+                            .upload_data(upload_request)
+                            .or(async {
+                                async_io::Timer::after(Duration::from_millis(1000)).await;
+                                Err(AppClientError::AppClientRequestTimeout)
+                            })
+                            .await
+                        {
+                            Ok(_) => {
+                                current_chunk_size = next_message.len();
+                                current_chunk.push(next_message);
+                                break true;
+                            }
+
+                            // If the request takes too long to elicit a response, we don't know whether
+                            // the upload was successful on app's side. We've decided that we'd rather
+                            // potentially lose some data than upload duplicate data, and so we opt to
+                            // assume the best and consume the messages
+                            Err(AppClientError::AppClientRequestTimeout) => {
+                                current_chunk_size = next_message.len();
+                                current_chunk.push(next_message);
+                                break true;
+                            }
+                            Err(err) => return Err(err),
```

Comment:
If we hit this case on the Nth chunk but we have actually sent the prior N-1 chunks OK, doesn't this direct exit from the function mean we will not flush? So those other chunks will get resent on the next cycle?

Reply:
I don't think that's true? The inner loop (L. 300) represents the processing of single chunk and the outer loop (L. 257) is looping over the region one chunk at a time. Since the flush is happening at the end of an iteration of the outer loop, it is happening at the end of each single chunk processed. I did this to prompt each chunk to have its own reader and so that returning here would exit the entire sync loop (so we don't keep requesting when the network is down).

This confusion is a definite indicator that I need to refactor this code for clarity

Reply:
Having re-read it, you are correct!